### PR TITLE
Fixes calculation of average thread blocked time and average thread waited time

### LINKED ISSUE #92: [BUG] Thread_Blocked_Time metric calculation is incorrect in Metrics API
**What is the bug?**
When a thread is blocked continuously for a prolonged period of time, its contribution to the average Thread_Blocked_Time metric is expected to increase; but the metrics API call keeps returning a constant value of 0.5 for the thread_blocked_time metric aggregated by average.

https://github.com/opensearch-project/performance-analyzer-rca/blob/eeb537fb120b2f79bf8d2cc4cb9cbabf75b06d3f/src/main/java/org/opensearch/performanceanalyzer/jvm/ThreadList.java#L282-L293

To calculate the average thread_blocked_time of a thread, the code first _checks if the thread's blocked_count from the current interval's threaddump is greater than the thread's blocked_count from the previous interval's threaddump_

The problem with this check is that if the thread has simply been blocked on the same monitor across both intervals, the blocked count doesn't increase but the blocked time would have increased by an amount equal to the threadumps interval duration. Hence the thread_blocked_time value is calculated by a default fallback logic rather than showing the  the increase in blocked time.  

**How can one reproduce the bug?**
**Thread dumps** for a cluster shows one of the many blocked search threads:

> "opensearch[0fd2634cfd28ef09d9924ce73c3a64b1][search][T#1]" #169 daemon prio=5 os_prio=0 cpu=59.36ms elapsed=644.32s tid=0x00007f91040dd000 nid=0x6172 waiting for monitor entry  [0x00007f90d07e1000]
>    java.lang.Thread.State: BLOCKED (on object monitor)
>         - waiting to lock <0x000000070e6fea30> (a java.lang.Object)
>         at org.opensearch.action.search.SearchProgressListener.notifyQueryResult(SearchProgressListener.java:135)
>         at org.opensearch.action.search.QueryPhaseResultConsumer.consumeResult(QueryPhaseResultConsumer.java:127)
>         at org.opensearch.action.search.AbstractSearchAsyncAction.onShardResult(AbstractSearchAsyncAction.java:486)
>         at org.opensearch.action.search.SearchQueryThenFetchAsyncAction.onShardResult(SearchQueryThenFetchAsyncAction.java:120)
>         at org.opensearch.action.search.AbstractSearchAsyncAction$1.innerOnResponse(AbstractSearchAsyncAction.java:261)
>         at org.opensearch.action.search.SearchActionListener.onResponse(SearchActionListener.java:58)
>         at org.opensearch.action.search.SearchActionListener.onResponse(SearchActionListener.java:42)
>         at org.opensearch.action.search.SearchExecutionStatsCollector.onResponse(SearchExecutionStatsCollector.java:81)
>         at org.opensearch.action.search.SearchExecutionStatsCollector.onResponse(SearchExecutionStatsCollector.java:49)
>         at org.opensearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:67)
>         at org.opensearch.action.search.SearchTransportService$ConnectionCountingHandler.handleResponse(SearchTransportService.java:413)
>         at org.opensearch.transport.TransportService$6.handleResponse(TransportService.java:658)
>         at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1207)
>         at org.opensearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:1285)
>         at org.opensearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1265)
>         at org.opensearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:65)
>         at org.opensearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:56)
>         at org.opensearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:40)
>         at org.opensearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:71)
>         at org.opensearch.action.ActionRunnable$$Lambda$3388/0x0000000801044040.accept(Unknown Source)
>         at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:86)
>         at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:50)
>         at org.opensearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:57)
>         at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:765)
>         at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:50)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.6/ThreadPoolExecutor.java:1128)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.6/ThreadPoolExecutor.java:628)
>         at java.lang.Thread.run(java.base@11.0.6/Thread.java:834)

**PA metrics api** call for ThreadBlockedTime metric with operation as dimension
`curl "http://localhost:9650/_opendistro/_performanceanalyzer/metrics?metrics=Thread_Blocked_Time&agg=avg&dim=Operation"`
`{"hLHH2jmQSrupS4MSyictKg": {"timestamp": 1630994665000, "data": {"fields":[{"name":"Operation","type":"VARCHAR"},{"name":"Thread_Blocked_Time","type":"DOUBLE"}],"records":[["GC",0.0],["flush",0.0],["generic",0.0],["get",0.0],["management",0.0],["other",0.0],["refresh",0.0],["search",0.5]]}}}`

The **thread_blocked_time** value for the search threads is shown as 0.5 while multiple threads had been blocked for atleast a minute.

**What is the expected behavior?**
The thread_blocked_time metric in Metrics API should more accurately account for average thread blocked time.

### PR DESCRIPTION
**Pull request by eirsep**
- Enables thread contention monitoring. 
- Fixes calculation of average thread blocked time and average thread waited time when thread is blocked/waiting due to same event across thread dump intervals.

Signed-off-by: Surya Sashank Nistala <snistala@amazon.com>

### Issues resolved
#92  

### Testing
```
for i in {1..3};do curl "http://localhost:9600/_plugins/_performanceanalyzer/metrics?metrics=Thread_Blocked_Time,Thread_Blocked_Event&agg=max,max&dim=Operation" ;echo;sleep 5;done
{"k1yr46alRW2mC5h7nCSSYQ": {"timestamp": 1643018630000, "data": {"fields":[{"name":"Operation","type":"VARCHAR"},{"name":"Thread_Blocked_Time","type":"DOUBLE"},{"name":"Thread_Blocked_Event","type":"DOUBLE"}],"records":[["GC",0.0,0.0],["flush",0.0,0.0],["generic",0.0,149.0],["management",0.0,38.0],["other",10.015115325794273,27.0],["refresh",0.0,0.0],["search",10.015115302620272,24.0],["transportWorker",0.0,87.0]]}}}

{"k1yr46alRW2mC5h7nCSSYQ": {"timestamp": 1643018635000, "data": {"fields":[{"name":"Operation","type":"VARCHAR"},{"name":"Thread_Blocked_Time","type":"DOUBLE"},{"name":"Thread_Blocked_Event","type":"DOUBLE"}],"records":[["GC",0.0,0.0],["flush",0.0,0.0],["generic",0.0,149.0],["management",0.0,38.0],["other",10.013015012003,27.0],["refresh",0.0,0.0],["search",10.013015012002999,24.0],["shardquery",3.112,1.0],["transportWorker",0.0,87.0]]}}}

{"k1yr46alRW2mC5h7nCSSYQ": {"timestamp": 1643018640000, "data": {"fields":[{"name":"Operation","type":"VARCHAR"},{"name":"Thread_Blocked_Time","type":"DOUBLE"},{"name":"Thread_Blocked_Event","type":"DOUBLE"}],"records":[["GC",0.0,0.0],["flush",0.0,0.0],["generic",0.0,149.0],["management",0.0,38.0],["other",10.013238956117647,27.0],["refresh",0.0,0.0],["search",10.01323896387822,24.0],["transportWorker",0.0,87.0]]}}}

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/performance-analyzer-rca/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: null/data/batch_metrics_enabled.conf [hunk 1]
```diff
@@ -0,0 +1 @@
+true
```
Comment (sruti1312) [inline null/data/batch_metrics_enabled.conf:1]: Can we remove these ```null/*/*.conf``` files? They are generated during test execution. Maybe adding ```null/*``` to ```.gitignore``` will help avoid these issues in the future.

File: null/data/thread_contention_monitoring_enabled.conf [hunk 2]
```diff
@@ -0,0 +1 @@
+true
```
Comment (sruti1312) [inline null/data/thread_contention_monitoring_enabled.conf:1]: same as above

File: src/main/java/org/opensearch/performanceanalyzer/jvm/ThreadList.java [hunk 3]
```diff
@@ -138,8 +138,13 @@ public String toString() {
...
      *
      * @return A hashmap of threadId to threadState.
+     * @param threadContentionMonitoringEnabled
      */
-    public static Map<Long, ThreadState> getNativeTidMap() {
+    public static Map<Long, ThreadState> getNativeTidMap(
+            boolean threadContentionMonitoringEnabled) {
+        if (threadBean.isThreadContentionMonitoringSupported()) {
+            threadBean.setThreadContentionMonitoringEnabled(threadContentionMonitoringEnabled);
+        }
```
Comment (sruti1312) [inline src/main/java/org/opensearch/performanceanalyzer/jvm/ThreadList.java:147]: This check and set happens during every ```OSMetricsCollector#collectMetrics``` call which is very frequent. Do we want to set it once during initialization and when there is state change in ```threadContentionMonitoringEnabled```?

**Review by sruti1312 — CHANGES_REQUESTED** [review]

**Comment by sruti1312** [comment]
Flaky ```EventLogFileHandlerTests``` test has been fixed. Currently, Integration tests (spin up a docker cluster and run a bunch of tests) are failing due to missing metrics (related to OSMetricsCollector). @eirsep Can you look into the failure? Thanks!

File: src/main/java/org/opensearch/performanceanalyzer/jvm/ThreadList.java (hunk 3 above)
Comment (eirsep) [reply src/main/java/org/opensearch/performanceanalyzer/jvm/ThreadList.java:147]: This check is not very costly
If the flag is already enabled then this short-circuits out of the method without doing anything.
 ```
if (contentionMonitoringEnabled != enable) {
                if (enable) {
                    // if reeabled, reset contention time statistics
                    // for all threads
                    resetContentionTimes0(0);
                }

                // update the VM of the state change
                setThreadContentionMonitoringEnabled0(enable);

                contentionMonitoringEnabled = enable;
            }
```

Also, the OSMetricsCollector is initialized in the PA Plugin. We can't do a set once and update on state change sort of thing from there as ThreadList.getNativeTidMap is a static call and will need to know thread contention monitoring is enabled or not before every run.

File: null/data/batch_metrics_enabled.conf (hunk 1 above)
Comment (eirsep) [reply null/data/batch_metrics_enabled.conf:1]: ack

**Comment by sruti1312** [comment]
Integration tests are failing in main as well.
https://github.com/opensearch-project/performance-analyzer-rca/runs/5842874823?check_suite_focus=true

**Review by sruti1312 — DISMISSED** [review]

**Review by sgup432 — DISMISSED** [review]

**Comment by sgup432** [comment]
Approving. Integ test failing due to known failure which was fixed recently and seems like this PR didn't pull that latest commit.

**Review by sgup432 — APPROVED** [review]

**Review by sruti1312 — APPROVED** [review]

**Pull request merged**